### PR TITLE
`@MustCallAlias` annotation must appear in both return type and parameter type (or not at all)

### LINKED ISSUE #6376: RLC: warn when `@MustCallAlias` is written on a return type but no parameter type, and vice versa
I don't think that we have any special handling for this case. Because `@MustCallAlias` is a poly annotation, we do get warnings sometimes (if it is placed in a way that is inconsistent with the general poly rules), but (as is clear from this example) not always.

I agree that it would be good for the RLC to warn if there is an `@MustCallAlias` annotation on a parameter but not a return and vice-versa, because that's the only reasonable way to use the annotation.

_Originally posted by @kelloggm in https://github.com/typetools/checker-framework/pull/6375#discussion_r1430295599_

### PR DESCRIPTION
**Pull request by jyoo980**
Closes #6376

### CODE REVIEW

File: checker/tests/resourceleak/MustCallAliasPassthroughWrong3.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 class MustCallAliasPassthroughWrong3 {
 
+  // :: warning: (mustcallalias.method.return.and.param)
```
Comment (jyoo980) [inline checker/tests/resourceleak/MustCallAliasPassthroughWrong3.java:9]: Should this warning be added, _or_ should the return type be annotated with `@MustCallAlias`?

Comment (kelloggm) [reply checker/tests/resourceleak/MustCallAliasPassthroughWrong3.java:9]: This test case was intended to cover the case when a user forgets an `@MustCallAlias` annotation on an otherwise correct implementation of a wrapper type. So, I think adding the warning is the right choice.

**Comment by jyoo980** [comment]
@kelloggm Do you have any insight into some of the failing tests? They appear to be related to RLC inference. I have one test case here I can show. For the code below:

```java
  private class MCAConstructor extends Foo {

    protected final @Owning Foo f;
    protected long s = 0L;

    // The Must Call Checker for assigning @MustCallAlias parameters to @Owning fields reports a
    // false positive.
    @SuppressWarnings("assignment")
    protected MCAConstructor(Foo foo) throws IOException {
      if (foo == null) {
        this.s = foo.b();
      }
      this.f = foo;
    }
```

The parameter `foo` appears to have the annotation `@MustCallAlias`, but I am unsure where it could be coming from (or perhaps my code is incorrect). The code I wrote to see if a parameter is annotated with a `@MustCallAlias` annotation is below:

```java
  /**
   * Given a method and its parameter list, look through each of the parameters and see if any are
   * annotated with the {@code @MustCallAlias} annotation.
   *
   * <p>Return true if at least one parameter has the {@code @MustCallAlias} annotation, otherwise
   * return false.
   *
   * @param tree the method declaration.
   * @param mcAtf a MustCallAnnotatedTypeFactory.
   * @return true if and only if a {@code @MustCallAlias} annotation appears at least once in the
   *     parameter list.
   */
  private boolean isMustCallAliasAnnoPresentInParams(
      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
    for (VariableTree paramDecl : tree.getParameters()) {
      VariableElement paramElt = TreeUtils.elementFromDeclaration(paramDecl);
      if (mcAtf.getDeclAnnotation(paramElt, MustCallAlias.class) != null) {
        return true;
      }
    }
    return false;
  }
```

**Comment by kelloggm** [comment]
> Do you have any insight into some of the failing tests? They appear to be related to RLC inference

These are definitely inference tests. It looks like what is happening is that:
1. during the inference phase, an `@MustCallAlias` annotation is inferred on the parameter (this makes sense, because the parameter is assigned into an owning field of the containing class inside a constructor, so the parameter should be `@MustCallAlias` with the constructor result)
2. during the validation phase (when the inference tests check that all warnings are gone), your new code marks the inferred `@MustCallAlias` annotation as invalid.

There are two possibilities for what's going wrong here:
* there is something wrong in your validation logic for constructor results. Do you have a test case where a constructor's parameter and the constructor result are both `@MustCallAlias`? The answer is almost certainly yes - there should be one in the existing RLC test suite. I don't see a reason why your logic would fail only during inference testing, but I suppose I could be overlooking something.
* inference is producing the wrong result for these test cases, and we didn't notice. In particular, inference might be inferring `@MustCallAlias` on the *parameter*, but not on the constructor result. On `master`, it's currently behaving correctly: I ran `./gradlew ainferResourceLeakAjavaTest` and then inspected `tests/ainfer-resourceleak/inference-output/MustCallAliasOnRegularExits-org.checkerframework.checker.resourceleak.ResourceLeakChecker.ajava`, which has the inference results for the example that you included above; I observed that the RLC infers `@MustCallAlias` on both the constructor result and on the parameter.

You should investigate both of these possibilities. One thing you could do is run the inference tests locally and then copy the inference test result (i.e., the file I referenced in the last bullet) into the regular RLC tests, which might make it easier to debug.

**Comment by jyoo980** [comment]
I think l see the issue, I was using `AnnotationUtils.getExplicitAnnotationsOnConstructorResult` on the constructor; which I assume does _not_ return annotations that are inferred (as those are implicit).

**Comment by kelloggm** [comment]
> I think l see the issue, I was using AnnotationUtils.getExplicitAnnotationsOnConstructorResult on the constructor; which I assume does not return annotations that are inferred (as those are implicit).

Yes, that makes sense. The annotations on the constructor there are stored in an `.ajava` file, so you'll need to use an API for getting the annotations that also returns annotations in ajava/stub files.

File: checker/src/main/java/org/checkerframework/checker/mustcall/messages.properties [hunk 2]
```diff
@@ -1,3 +1,5 @@
 inconsistent.mustcall.subtype=%s is annotated as %s, but one of its supertypes is annotated %s
 createsmustcallfor.target.unparseable=The method %s is annotated as @CreatesMustCallFor, but the target (%s) was unparseable in the current context. Rewrite your code so that the relevant expression is a local variable.
 mustcall.not.inheritable=@MustCall annotation is not inherited by subclasses. Use @InheritableMustCall instead.
+mustcallalias.method.return.and.param=A @MustCallAlias annotation must appear on a return type and a parameter type
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/mustcall/messages.properties:4]: Let's add a string argument to this error message that indicates which location is missing an annotation. I suggest the following: 
```@MustCallAlias annotations must appear in pairs (one on a return type and one a parameter type), but %s```, with the following completions for the `%s`:
* `this method has a void return.`
* `there is no @MustCallAlias annotation on the return type, even though parameter $p is annotated as @MustCallAlias.`
* `no parameter has an @MustCallAlias annotation, even though the return type is @MustCallAlias.`

You'll have to distinguish between the last two in the code that calls `checker.reportError`, but I think that's worthwhile if it leads to more actionable error messages.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 3]
```diff
@@ -201,6 +209,86 @@ private void checkOwningOverrides(
...
+    if (TreeUtils.isVoidReturn(tree) && isMustCallAliasAnnoOnParameter) {
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param");
+      return;
+    }
+
+    AnnotatedTypeMirror returnType = mcAtf.getMethodReturnType(tree);
+    boolean isMustCallAliasAnnoOnReturnType = returnType.hasPrimaryAnnotation(PolyMustCall.class);
+
+    if (isMustCallAliasAnnoOnParameter != isMustCallAliasAnnoOnReturnType) {
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param");
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:236]: If you take my suggestion about the warnings, this is where you'll need to make a corresponding change.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 4]
```diff
@@ -201,6 +209,86 @@ private void checkOwningOverrides(
...
+      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
+    ExecutableElement constructorDecl = TreeUtils.elementFromDeclaration(tree);
+    boolean isMustCallAliasAnnoOnConstructor =
+        constructorDecl != null
+            && constructorDecl.getAnnotationMirrors().stream()
+                .anyMatch(anno -> atypeFactory.areSameByClass(anno, MustCallAlias.class));
+    boolean isMustCallAliasAnnoOnParam = isMustCallAliasAnnoPresentInParams(tree, mcAtf);
+
+    if (isMustCallAliasAnnoOnConstructor != isMustCallAliasAnnoOnParam) {
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param");
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:260]: A change here will also be needed.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 5]
```diff
@@ -201,6 +209,86 @@ private void checkOwningOverrides(
...
+   * @param tree the constructor.
+   * @param mcAtf a MustCallAnnotatedTypeFactory.
+   */
+  private void checkMustCallAliasAnnotationForConstructor(
+      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
+    ExecutableElement constructorDecl = TreeUtils.elementFromDeclaration(tree);
+    boolean isMustCallAliasAnnoOnConstructor =
+        constructorDecl != null
+            && constructorDecl.getAnnotationMirrors().stream()
+                .anyMatch(anno -> atypeFactory.areSameByClass(anno, MustCallAlias.class));
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:256]: Rather than this implementation, I think you could instead reuse `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)`, whose implementation is:
```
  /*package-private*/ boolean hasMustCallAlias(Element elt) {
    if (noResourceAliases) {
      return false;
    }
    MustCallAnnotatedTypeFactory mustCallAnnotatedTypeFactory =
        getTypeFactoryOfSubchecker(MustCallChecker.class);
    return mustCallAnnotatedTypeFactory.getDeclAnnotationNoAliases(elt, MustCallAlias.class)
        != null;
  }
```

I don't think that's directly accessible from `MustCallVisitor`, but I think it would be okay to move the implementation into the Must Call Checker and then have the existing method call a delegate.

**Review by kelloggm — CHANGES_REQUESTED** [review]

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java (hunk 5 above)
Comment (jyoo980) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:256]: I'm a bit confused here: the code you commented on is implemented within the `ResourceLeakVisitor`, and not the `MustCallChecker`:

> I think it would be okay to move the implementation into the Must Call Checker and then have the existing method call a delegate.

Do you mean I should move the code I've written (i.e., to check that the `@MustCallAlias` annotation occurs in pairs) into the Must Call Checker? Or did you mean for me to move the implementation of the `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)` into the Must Call Checker?

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:256]: > I'm a bit confused here: the code you commented on is implemented within the ResourceLeakVisitor, and not the MustCallChecker:

This was my mistake - I thought I was reading `MustCallVisitor`. You can disregard that part of the comment.

> I think `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)` is accessible from the `ResourceLeakVisitor`

It is, and I think you should call it instead of writing new code that does the same job. Sorry for the confusion!

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 6]
```diff
@@ -201,6 +209,93 @@ private void checkOwningOverrides(
...
+
+    if (isMustCallAliasAnnoOnConstructor != isMustCallAliasAnnoOnParameter) {
+      String message =
+          isMustCallAliasAnnoOnParameter
+              ? String.format(
+                  "there is no @MustCallAlias annotation on the return type, even though parameter %s is annotated with @MustCallAlias",
+                  paramWithMustCallAliasAnno)
+              : "no parameter has a @MustCallAlias annotation, even though the return type is annotated with @MustCallAlias";
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param", message);
+    }
```
Comment (jyoo980) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:272]: I don't particularly like this duplicated code (same code in `checkMustCallAliasAnnotationForMethod`) but I'm not sure if extracting it into its own method would be overkill here

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:272]: I think it's fine to factor it out. I almost always prefer helper methods for this kind of thing.

**Comment by jyoo980** [comment]
@kelloggm You suggested the following message string:

> there is no @MustCallAlias annotation on the return type, even though parameter $p is annotated as @MustCallAlias

I'm not sure if this message will be accurate for Java constructors; they don't have a return type. Is there a way to make this more general? We could also have another message string specific to constructors.

**Comment by kelloggm** [comment]
> I'm not sure if this message will be accurate for Java constructors; they don't have a return type. Is there a way to make this more general? We could also have another message string specific to constructors.

Fair point. If you can reword it to include constructors that's ok with me, but it's also fine to have another message. My goal with that comment was to nudge you towards a set of messages that will be more helpful to the user.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 7]
```diff
@@ -201,6 +209,108 @@ private void checkOwningOverrides(
...
+   * @param isMustCallAliasAnnoOnMethodOrConstructorDecl true if and only if a
+   *     {@code @MustCallAlias} annotation appears on a method or constructor declaration
+   * @param tree the method or constructor declaration
+   */
+  private void checkMustCallAliasAnnoMismatch(
+      @Nullable Element paramWithMustCallAliasAnno,
+      boolean isMustCallAliasAnnoOnMethodOrConstructorDecl,
+      MethodTree tree) {
+    boolean isMustCallAliasAnnotationOnParameter = paramWithMustCallAliasAnno != null;
+    String locationOfCheck = TreeUtils.isClassTree(tree) ? "this constructor" : "the return type";
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:278]: Move this computation inside the `if` statement, since it is not used until there.

**Review by kelloggm — APPROVED** [review]
I don't need to review again.

**Comment by jyoo980** [comment]
@kelloggm sorry for the ping, but it looks like my latest push to address comments made your approval stale.

@msridhar @smillst I would appreciate a review if you're interested!

**Review by msridhar — APPROVED** [review]
LGTM, thanks for this new check!
